File: edukate-notifier/src/main/java/io/github/sanyavertolet/edukate/notifier/dtos/BaseNotificationDto.java
package io.github.sanyavertolet.edukate.notifier.dtos;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import lombok.AllArgsConstructor;
import lombok.Getter;

import java.time.Instant;

@JsonTypeInfo(
        use = JsonTypeInfo.Id.NAME,
        include = JsonTypeInfo.As.PROPERTY,
        property = "_type",
        visible = true
)
@JsonSubTypes({
        @JsonSubTypes.Type(value = BaseNotificationDto.class, name = "base"),
        @JsonSubTypes.Type(value = SimpleNotificationDto.class, name = "simple"),
        @JsonSubTypes.Type(value = InviteNotificationDto.class, name = "invite"),
        @JsonSubTypes.Type(value = CheckedNotificationDto.class, name = "checked"),
})
@Getter
@AllArgsConstructor
public sealed class BaseNotificationDto permits
        SimpleNotificationDto,
        InviteNotificationDto,
        CheckedNotificationDto {
    private String uuid;
    private Boolean isRead;
    private Instant createdAt;
}

File: edukate-notifier/src/main/kotlin/io/github/sanyavertolet/edukate/notifier/dtos/BaseNotificationDto.kt
package io.github.sanyavertolet.edukate.notifier.dtos

import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonPropertyOrder
import com.fasterxml.jackson.annotation.JsonSubTypes
import com.fasterxml.jackson.annotation.JsonTypeInfo
import java.time.Instant

@JsonPropertyOrder("createdAt", "uuid", "isRead", "_type")
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "_type")
@JsonSubTypes(
    JsonSubTypes.Type(value = SimpleNotificationDto::class, name = "simple"),
    JsonSubTypes.Type(value = InviteNotificationDto::class, name = "invite"),
    JsonSubTypes.Type(value = CheckedNotificationDto::class, name = "checked"),
)
sealed interface BaseNotificationDto {
    val uuid: String

    @get:JsonProperty("isRead") val isRead: Boolean
    val createdAt: Instant
}
